### PR TITLE
IChatCompletion Multiple Results / New Interfaces

### PR DESCRIPTION
**Pull request by rogerbarreto**
### Motivation and Context

Add potential for IChatCompletion with multiple results per prompt.

MultiChatResultPR

### Description

- Adds new Interfaces/Abstractions for Chat Results: `IChatResult`, `IChatStreamingResult`, `IChatMessage`
- Adds new `IChatCompletion` interface methods for Streaming and Non-Streaming results
- Moves old `IChatCompletion` interface methods `IAsyncEnumerable<string> GenerateMessageStreamAsync` and `Task<string> GenerateMessageAsync` as extensions for retro compatibility
- Changed `ChatAsText` to use the new multiple results implementations
- Makes `ChatHistory.ChatMessage` & `ChatHistory.AuthorsRole` types obsolete (discouraging usage of those implementation in favor of new `IChatMessage`
- `ChatHistory` abstraction now extends from `List<IChatMessage>`
- Updates all Samples using the new interfaces. 

### Paves the way to ... (Futures)
- Remove `ChatHistory.ChatMessage` inner class
- Remove `ChatHistory.AuthorsRole` enums 
- Make `ChatHistory` an abstract class or an `IChatHistory` interface.
- Rename `ITextCompletionResult` to `ITextResult`
- Rename `ITextCompletionStreamingResult` to `ITextStreamingResult`

### Contribution Checklist
- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ChatCompletionResult.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+using System.Threading.Tasks;
+using Azure.AI.OpenAI;
+using Microsoft.SemanticKernel.AI.ChatCompletion;
+using Microsoft.SemanticKernel.AI.TextCompletion;
+using Microsoft.SemanticKernel.Diagnostics;
+using Microsoft.SemanticKernel.Orchestration;
+
+namespace Microsoft.SemanticKernel.Connectors.AI.OpenAI.AzureSdk;
+
+internal sealed class ChatResult : IChatResult, ITextCompletionResult
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ChatCompletionResult.cs:13]: Class name (`ChatResult`) doesn't match filename (`ChatCompletionResult`).

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ChatCompletionStreamingResult.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+using Azure.AI.OpenAI;
+using Microsoft.SemanticKernel.AI;
+using Microsoft.SemanticKernel.AI.ChatCompletion;
+using Microsoft.SemanticKernel.AI.TextCompletion;
+using Microsoft.SemanticKernel.Diagnostics;
+using Microsoft.SemanticKernel.Orchestration;
+
+namespace Microsoft.SemanticKernel.Connectors.AI.OpenAI.AzureSdk;
+
+internal sealed class ChatStreamingResult : IChatStreamingResult, ITextCompletionStreamingResult
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ChatCompletionStreamingResult.cs:17]: Class name (`ChatStreamingResult`) doesn't match filename (`ChatCompletionStreamingResult`).

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs [hunk 3]
```diff
@@ -135,22 +135,22 @@ await foreach (StreamingChoice choice in streamingChatCompletions.GetChoicesStre
...
-    private protected async Task<string> InternalGenerateChatMessageAsync(
+    private protected async Task<IReadOnlyList<IChatResult>> InternalGetChatResultsAsync(
         ChatHistory chat,
-        ChatRequestSettings requestSettings,
+        ChatRequestSettings? chatSettings,
         CancellationToken cancellationToken = default)
     {
         Verify.NotNull(chat);
-        Verify.NotNull(requestSettings);
+        chatSettings ??= new ChatRequestSettings();
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:147]: Small simplification as in other places 😄 
```suggestion
        chatSettings ??= new();
```

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs [hunk 4]
```diff
@@ -284,10 +261,8 @@ private static ChatHistory PrepareChatHistory(string text, CompleteRequestSettin
 
     private static CompletionsOptions CreateCompletionsOptions(string text, CompleteRequestSettings requestSettings)
     {
-        if (requestSettings.ResultsPerPrompt < 1 ||
-            requestSettings.ResultsPerPrompt > 128)
+        if (requestSettings.ResultsPerPrompt is < 1 or > 128)
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:264]: It's not clear what is 128 here. Let's put these numbers in separate constant variables with names.

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs [hunk 5]
```diff
@@ -317,16 +292,21 @@ private static CompletionsOptions CreateCompletionsOptions(string text, Complete
         return options;
     }
 
-    private static ChatCompletionsOptions CreateChatCompletionsOptions(ChatRequestSettings requestSettings, ChatHistory chat)
+    private static ChatCompletionsOptions CreateChatCompletionsOptions(ChatRequestSettings requestSettings, IEnumerable<ChatMessageBase> chatHistory)
     {
+        if (requestSettings.ResultsPerPrompt is < 1 or > 128)
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:297]: It's not clear what is 128 here. Let's put these numbers in separate constant variables with names.

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs [hunk 6]
```diff
@@ -337,22 +317,28 @@ private static ChatCompletionsOptions CreateChatCompletionsOptions(ChatRequestSe
...
+            ValidateChatAuthors(message.Role, out var validRole);
 
-            options.Messages.Add(new ChatMessage(role, message.Content));
+            options.Messages.Add(new ChatMessage(validRole, message.Content));
         }
 
         return options;
     }
 
+    private static void ValidateChatAuthors(AuthorRole role, out ChatRole validRole)
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:330]: Is it necessary to use `out` parameter here? Can we make it simpler like:
```csharp
private static ChatRole GetValidChatRole(AuthorRole role)
```

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs [hunk 7]
```diff
@@ -337,22 +317,28 @@ private static ChatCompletionsOptions CreateChatCompletionsOptions(ChatRequestSe
...
+    {
+        validRole = new ChatRole(role.Label);
+
+        if (validRole != ChatRole.User &&
+            validRole != ChatRole.System &&
+            validRole != ChatRole.Assistant)
+        {
+            throw new ArgumentException($"Invalid chat message author: {role}");
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:338]: Small improvement in exception message:
```suggestion
            throw new ArgumentException($"Invalid chat message author role: {role}");
```

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/SKChatMessage.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using Microsoft.SemanticKernel.AI.ChatCompletion;
+
+namespace Microsoft.SemanticKernel.Connectors.AI.OpenAI.AzureSdk;
+
+public class SKChatMessage : ChatMessageBase
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/SKChatMessage.cs:7]: Missing XML documentation in public class and constructor.

File: dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistory.cs [hunk 9]
```diff
@@ -1,11 +1,15 @@
...
 
+using System;
 using System.Collections.Generic;
 
+#pragma warning disable CA1710
+
 namespace Microsoft.SemanticKernel.AI.ChatCompletion;
 
-public class ChatHistory
+public class ChatHistory : List<ChatMessageBase>
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistory.cs:10]: Missing XML documentation in public class and members.

File: dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistory.cs [hunk 10]
```diff
@@ -14,53 +18,85 @@ public enum AuthorRoles
...
-        /// </summary>
-        public string Content { get; set; }
-
         /// <summary>
         /// Create a new instance
         /// </summary>
         /// <param name="authorRole">Role of message author</param>
         /// <param name="content">Message content</param>
-        public Message(AuthorRoles authorRole, string content)
+        public Message(AuthorRoles authorRole, string content) : base(new AuthorRole(authorRole.ToString()), content)
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistory.cs:44]: `AuthorRoles` enum has 4 values `Unknown`, `System`, `User`, `Assistant`, starting with uppercase letter. `.ToString()` method of enum value will return the same string with first uppercase letter. 

While your new struct `AuthorRole` has 3 pre-configured roles with labels starting with lowercase letter: `system`, `assistant`, `user`. It looks like new approach is not consistent with old approach. 

Maybe from OpenAI API perspective it's not a problem, just want to make sure it won't introduce any bugs related to case sensitivity.

File: dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatMessageBase.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    /// <summary>
+    /// Creates a new instance of the <see cref="ChatMessageBase"/> class
+    /// </summary>
+    /// <param name="Role">Role of the author of the message</param>
+    /// <param name="Content">Content of the message</param>
+    protected ChatMessageBase(AuthorRole Role, string Content)
+    {
+        this.Role = Role;
+        this.Content = Content;
+    }
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatMessageBase.cs:29]: camelCase for constructor parameters?
```suggestion
    /// <summary>
    /// Creates a new instance of the <see cref="ChatMessageBase"/> class
    /// </summary>
    /// <param name="role">Role of the author of the message</param>
    /// <param name="content">Content of the message</param>
    protected ChatMessageBase(AuthorRole role, string content)
    {
        this.Role = role;
        this.Content = content;
    }
```

File: dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistory.cs (hunk 10 above)
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistory.cs:44]: Good point, I will move those to be capitalized as the old approach.

**Pull request closed without merging**